File: deconv_moffat_dcv/dcv_deconv_2D.f90
!****************************************************************** 
! Program in Fortran 90
!
!  spdiv => Direct spectral division
!  wiene => Wiener filter 
!  tikho => Tikhonov regularisation
!  maxen => Maximum Entropy method
!
! JLP
! Version 03/04/2001
!****************************************************************** 
program dcv_deconv_2D
 use dcv_2D_data
 use dcv_2D_mod
 implicit none
 real			:: rms_err,ss,L_phi,L_y,snr
 integer		:: nx0,ny0,iter,pntr
 character(len=1)	:: ans
 character(len=20)	:: prefix
 character(len=60)	:: filename
 character(len=80)	:: comments

call jlp_begin
call jlp_inquifmt

print *,' Program to deconvolve 2D signals'
print *,' JLP version 28/03/2001'

!!!! idim = 128*128 = 16384
!  prefix='s3'
  prefix='s5'

if(simulation)then
  print *,' Simulation: *_psf,*_yy0 are needed (prefix: s3, s5, ...)'
else
  print *,' Real data: *_psf,*_raw are needed'
endif
  print *,' Prefix for filenames: '
  read(5,*) prefix 
  print *,'OK: prefix:',prefix

! Read a file first to determine the size:
if(simulation)then
  filename=trim(prefix)//'_yy0'
else
  filename=trim(prefix)//'_raw'
endif
call jlp_vm_readimag(pntr,nx,ny,filename,comments);
!
! nn = number of pixels of an image (used by frprmn)
nn = nx*ny;
if(nn /= idim)then
  print *,' Fatal error: size of images is not equal to idim=',idim
  stop
endif
! To prepare FFT's:
if(nn > idim)then
  print *,' Fatal error: size of images larger than idim=',idim
  stop
endif
! To prepare FFT's:
print *,' fft_setup with ',nx,ny
call fft_setup(nx,ny)

! Then allocate memory space:
allocate(yy0(nn)); allocate(yy(nn)); allocate(hh(nn));
allocate(yyd(nn)); allocate(w0(nn)); allocate(w1(nn)); 
allocate(w_re(nn)); allocate(w_im(nn));
allocate(hh_re(nn)); allocate(hh_im(nn));
allocate(hht_re(nn)); allocate(hht_im(nn));
allocate(yy_re(nn)); allocate(yy_im(nn));

! Read input files:
!----------------------
if(simulation)then
! Original signal:
  filename=trim(prefix)//'_yy0'
  call jlp_readimag(yy0,nx,ny,nx,filename,comments);
else
! Image to deconvolve 
  filename=trim(prefix)//'_raw'
  call jlp_readimag(yy,nx,ny,nx,filename,comments);
endif

!----------------------
! PSF:
filename=trim(prefix)//'_psf'
call jlp_readimag(hh,nx0,ny0,nx,filename,comments);
! Normalize PSF (to avoid numerical pb...)
hh = hh / maxval(hh)
if(nx /= nx0 .or. ny /= ny0)then 
   print *,' Fatal error: wrong size for psf : nx=',nx0,' ny=',ny0; stop;
endif
! Compute transfer function:
  hh_re = hh; hh_im =0.;
  call fft_float(hh_re,hh_im,nx,ny,1)

!----------------------
! Signal to deconvolve:
if(simulation)then
! Either input it from a file
  if(.false.)then
! Original signal:
    filename=trim(prefix)//'_yyb'
    call jlp_readimag(yy,nx0,ny0,nx,filename,comments);
    if(nx /= nx0 .or. ny /= ny0)then 
       print *,' Fatal error: wrong size for yyb : nx=',nx0,' ny=',ny0; stop;
    endif
  else
! Or create it:  
! snr is in dB
! snr=20 or 13dB
! snr=100 or 20dB : simu 3: with disk
! snr=1000 or 30dB: simu 4: with disk
! snr=20 or 13dB: simu 5: without disk
    snr = 13
    call noisy_signal(snr)
    filename =trim(prefix)//'_yyb'
    write(comments,55) trim(prefix)//': convolution of yy0 with psf, SNR=',snr
55  format(A,F6.2,' dB')
    call jlp_writeimag(yy,nx,ny,nx,filename,comments);
  endif
! End simulation case
endif

! Set to zero all arrays:
w0 = 0.; w1 = 0.; w_re = 0.; w_im = 0.; 

!--------------------------------
! Deconvolution
!--------------------------------

write(6,18); 
18 format(/,&
' ==================== Options: ================',/,&
' - spdiv: spectral division',/,&
' - wiene: Wiener filter',/,&
' Regularisation on x_i only: ',/,&
' - tikho: Tikhonov''s regularisation',/,&
' - ggaus: Generalized Gauss''s regularisation (p=1.1)',/,&
' - maxen: regularisation with maximum entropy',/,&
' - sqrtr: convex sqrt(s2+x2) regularisation',/,&
' Regularisation on (x_i+1 - x_i): ',/,&
' - gmark: Gauss-Markov''s regularisation (p=2)',/,&
' ======== Enter the option you want: ===========')
read(5,*) option
print *,' OK: option=',option
L_phi = 0.; 

positivity = .false.
if(option(1:5) .ne. 'spdiv' .and. option(1:5) .ne. 'wiene'&
   .and. option(1:5) .ne. 'maxen')then
   print *,' Positivity constraint? [y/n]'; read(5,*)ans
   if((ans(1:1) .eq. 'y') .or. (ans(1:1) .eq. 'Y')) positivity=.true.
endif
if(option(1:5) .ne. 'spdiv' .and. option(1:5) .ne. 'wiene')then
   print *,' Enter alpha value: '; read(5,*)alpha
endif

select case(option)

!--------------------------------
! Deconvolution by spectral division:
case('spdiv')
   call dcv_spdiv
   write(comments,10)
10 format('Spectral division:')

!--------------------------------
! Deconvolution by Wiener filter:
case('wiene')
! alpha is the ratio variance of noise / variance noise of signal 
   if(simulation)then
     alpha=10**(-snr/10.)
     print *,'Theoretical value is (sigma_noise/sigma_signal)**2 :',alpha
   endif
   print *,' Enter alpha value: '; read(5,*)alpha
   call dcv_wiener
   write(comments,11) alpha
11 format('Wiener: alpha=',F4.1)

!--------------------------------
! Deconvolution with Tikhonov's regularisation 
! Generalized Gauss with p=2
case('tikho')
   call dcv_tikhonov(iter)
   write(comments,12) alpha,iter
12 format('Tikhonov: alpha=',F4.1,' iter=',I3)
   L_phi = phi_tikho(yyd)

!--------------------------------
! Deconvolution with generalized Gauss' regularisation 
! with p=1.1
case('ggaus')
   call dcv_ggauss(iter)
   write(comments,13) alpha,iter
13 format('Gen. Gauss: alpha=',F4.1,' it=',I3)
   L_phi = phi_ggauss(yyd)

!--------------------------------
! Deconvolution with convex sqrt(s2+x2) regularisation 
case('sqrtr')
   print *,' sqrt(s^2+x^2)/Enter s value: '; read(5,*)ss
   call dcv_sqrtrf(ss,iter)
   write(comments,14) ss,alpha,iter
14 format('sqrt: s=',F5.3,' alpha=',F4.1,' it=',I3)
   L_phi = phi_sqrtrf(yyd)

!--------------------------------
! Deconvolution with Maximum Entropy method 
case('maxen')
   call dcv_mem(iter)
   write(comments,15) alpha,iter
15 format('MEM: alpha=',F4.1,' iter=',I3)
   L_phi = phi_mem(yyd)

!--------------------------------
! Deconvolution with Gauss Markov regularisation (on x_i+1 - x_i) 
case('gmark')
   call dcv_gmark(iter)
   write(comments,16) alpha,iter
16 format('Gauss-Markov alpha=',F4.1,' it=',I3)
   L_phi = phi_gmark(yyd)

!---------------------------------
case default
   print *,' Fatal: invalid option'
   stop
end select

if(simulation)then
   rms_err=sqrt(sum((yyd-yy0)**2))/float(nn)
   write(comments,22) trim(comments),rms_err
22 format(a,' rms=',e10.3)
endif

!----------------- L Curve:
! L_phi = phi(x)
! L_y = || y - H x ||^2
    if(L_phi /= 0.)then
!      alpha = 0.;
      L_y = func(yyd)
      write(6,33) L_phi,L_y 
33    format('L curve: L_phi=',G10.4,' L_y=',G10.4)
    endif

filename =trim(prefix)//'_yyd'
call recent_fft(yyd,yyd,nx,ny,nx)
call jlp_writeimag(yyd,nx,ny,nx,filename,comments);


call jlp_end

end program dcv_deconv_2D 
!----------------------------------------------------------------
! Generates a noisy signal with a given SNR
! snr: in dB 
!----------------------------------------------------------------
subroutine noisy_signal(snr)
 use dcv_2D_data
 use dcv_2D_mod
 implicit none
real			:: snr,noise_level
!
! dcv_conv_hh returns: yy = conv(hh,yy0)
  call dcv_conv_hh(yy0,hh_re,hh_im,yy)
  call recent_fft(yy,yy,nx,ny,nx)
! Generate noisy signal with SNR in dB:
! Multiply by 12, since variance of uniform random function is 1/12.
  noise_level = 12. * (sum(yy**2) * 10**(-snr/10.)) / float(nn)
  noise_level = sqrt(noise_level)
  write(6,'("Simulated signal with SNR=",f8.2,"dB => noise level:",f8.3)') snr,noise_level
  call random_seed; call random_number(w0);
! Center noise:
  w0 = w0 - 0.5;
!
  yy = yy + w0 * noise_level;
return
end

File: deconv_moffat_dcv/dcv_deconv_2D.c
/****************************************************************** 
* dcv_deconv_2D
* Deconvolution program
*
*  spdiv => Direct spectral division
*  wiene => Wiener filter 
*  tikho => Tikhonov regularisation
*  maxen => Maximum Entropy method
*
* JLP
* Version 17/04/2006
*******************************************************************/ 
#include "jlp_dcv.h"

extern double *yy0, *yy, *yyd;

int dcv_remove_background(double *hh, int nx, int ny, 
                          double *backg, double *noise);
int dcv_2D_get_param(double *yy, int nx, int ny, double *alpha, 
                     double *ss, int *positivity, double *ftol,
                     double noise, char *option);

int main(int argc, char *argv[])
{
double *hh, rms_err, ss, ssum, alpha, ftol; 
double maxval, backg, noise, snr, L_phi, L_y;
float *float_tab;
int simu, iter, positivity;
INT_PNTR pntr;
INT4 nx, ny, nx0, ny0;
register int i;
char prefix[20], filename[60], comments[80];
char option[12];

printf(" Program to deconvolve 2D signals\n");
printf(" JLP version 23/04/2002\n");

/* Reduce the number of arguments when "runs" is used and  argc=7 */
if(argc == 7)
 {
 for(i = 6; i > 0; i--)
  {
   if(*argv[i] == ' ' || *argv[i] == '\0')argc--;
   else break;
  }
 }

if(argc == 3)
{
sscanf(argv[1],"%d",&simu);
sscanf(argv[2],"%s",prefix);
}
else
{
printf(" Usage: dcv_deconv_2D simu prefix \n");
printf(" example: dcv_deconv_2D 1 s5 \n\n");
printf(" Simulation (simu=1): *_psf,*_yy0 are needed (prefix: s3, s5, ...)\n");
printf(" Real data (simu=0): *_psf,*_raw are needed\n");
exit(-1);
}

if(simu)
  printf(" OK: simulation\n");
else
  printf(" OK: real data\n");

JLP_INQUIFMT();

/* Read psf first to determine the size:
*/
sprintf(filename,"%s_psf",prefix);
JLP_VM_READIMAG1(&pntr, &nx, &ny, filename, comments);

/* Then allocate memory space: */
dcv_2D_init(nx, ny);

/**************** 
*  Read input files:
*****************/

/*----------------------
* Original signal:
*/
if(simu) {
   sprintf(filename,"%s_yy0",prefix);
   JLP_VM_READIMAG1(&pntr, &nx0, &ny0, filename, comments);
      if(ny != ny0 || nx != nx0) {
      printf(" Fatal error: incompatible size! (nx=%d,ny=%d) (nx0=%d, ny0=%d)\n", 
              nx, ny, nx0, ny0);
      exit(-1);
      }
   float_tab = (float *) pntr;
   TO_DOUBLE(float_tab, &yy0[1], &nx, &ny, &nx);
   } else {
/*----------------------
* Signal to deconvolve:
*/
   sprintf(filename,"%s_raw",prefix);
   JLP_VM_READIMAG1(&pntr, &nx0, &ny0, filename, comments);
      if(ny != ny0 || nx != nx0) {
      printf(" Fatal error: incompatible size! (nx=%d,ny=%d) (nx0=%d, ny0=%d)\n", 
              nx, ny, nx0, ny0);
      exit(-1);
      }
   float_tab = (float *) pntr;
   TO_DOUBLE(float_tab, &yy[1], &nx, &ny, &nx);
   }

/*----------------------
* PSF:
*/
sprintf(filename,"%s_psf",prefix);
JLP_VM_READIMAG1(&pntr, &nx, &ny, filename, comments);
float_tab = (float *) pntr;
if((hh = (double *)malloc((nx * ny + 1) * sizeof(double))) == NULL)
  {
  printf("fatal error allocating memory space for hh\n");
  exit(-1);
  }
TO_DOUBLE(float_tab, &hh[1], &nx, &ny, &nx);

/* Remove background from the psf: */
 dcv_remove_background(hh, nx, ny, &backg, &noise);

#ifdef DEBUG
      sprintf(filename,"%s_psf_flat",prefix);
      sprintf(comments,"psf with background removed (back=%.4e)", backg);
      JLP_D_WRITEIMAG(hh, &nx, &ny, &nx, filename, comments);
#endif

/* Normalize PSF (to avoid numerical pb...)
*/
maxval = hh[1];
for( i = 1; i <= nx * ny; i++) if(maxval < hh[i]) maxval = hh[i];
if(maxval == 0) { printf("Fatal error: PSF is null!\n"); exit(-1); }
for( i = 1; i <= nx * ny; i++) hh[i] /= maxval;


/* Compute transfer function:
*/
dcv_2D_transfert(hh);

/* Signal to deconvolve: */
if(simu) {
/* Either input it from a file */
    if(0) {
      sprintf(filename,"%s_yyb",prefix);
      JLP_VM_READIMAG1(&pntr, &nx0, &ny0, filename, comments);
        if(ny != ny0 || nx != nx0) {
        printf(" Fatal error: incompatible size! (nx=%d,ny=%d) (nx0=%d, ny0=%d)\n",
                nx, ny, nx0, ny0);
        exit(-1);
        }
      float_tab = (float *) pntr;
      TO_DOUBLE(float_tab, &yy[1], &nx, &ny, &nx);
    } else {
/* Or create it:
* snr is in dB
* snr=20 or 13dB
* snr=100 or 20dB : simu 3: with disk
* snr=1000 or 30dB: simu 4: with disk
* snr=20 or 13dB: simu 5: without disk
*/
      snr = 13.;
      dcv_2D_noisy_signal(snr);
      sprintf(filename,"%s_yyb",prefix);
      sprintf(comments,"%s: convolution of yy0 with psf, SNR=%.1f", prefix, snr);
      JLP_D_WRITEIMAG(yy, &nx, &ny, &nx, filename, comments);
    }
/* End of simulation case: */
}

/*--------------------------------
* Deconvolution
*--------------------------------
*/

printf(" ==================== Options: ================\n");
printf(" - spdiv: spectral division \n");
printf(" - wiene: Wiener filter \n");
printf(" Regularisation on x_i only \n");
printf(" - tikho: Tikhonov's regularisation\n");
printf(" - ggaus: Generalized Gauss's regularisation (p=1.1)\n");
printf(" - maxen: regularisation with maximum entropy\n");
printf(" - sqrtr: convex sqrt(s2+x2) regularisation\n");
printf(" Regularisation on x_i+1 - x_i \n");
printf(" - gmark: Gauss-Markov''s regularisation (p=2)\n");
printf(" ======== Enter the option you want: ===========\n");
scanf("%s", option);
printf(" OK: option= %s \n",option);

/* Remove background from the input signal: */
 dcv_remove_background(yy, nx, ny, &backg, &noise);
#ifdef DEBUG
      sprintf(filename,"%s_raw_flat",prefix);
      sprintf(comments,"with background removed (back=%.4e)", backg);
      JLP_D_WRITEIMAG(yy, &nx, &ny, &nx, filename, comments);
#endif
/* Obtain parameters for minimization: */
   dcv_2D_get_param(yy, nx, ny, &alpha, &ss, &positivity, &ftol,
                    noise, option); 

L_phi = 0.;
/*--------------------------------
* Deconvolution by spectral division:
*/
  if(!strncmp(option,"spdiv",5)) {
      strcpy(comments," Spectral division");
      dcv_2D_spdiv();
      }
/*--------------------------------
* Deconvolution by Wiener filter:
*/
  else if(!strncmp(option,"wiene",5)) {
      dcv_2D_wiener(alpha);
      sprintf(comments," Wiener: alpha=%.1f",alpha);
      }
/*--------------------------------
* Deconvolution with Tikhonov's regularisation 
* Generalized Gauss with p=2
*/
  else if(!strncmp(option,"tikho",5)) {
      dcv_2D_tikhonov(alpha, positivity, ftol, &iter, &L_phi, &L_y);
      sprintf(comments," Wiener: alpha=%.1f, iter=%d", alpha, iter);
      }
/*--------------------------------
* Deconvolution with generalized Gauss' regularisation 
* with p=1.1
*/
  else if(!strncmp(option,"ggaus",5)) {
      dcv_2D_ggauss(alpha, positivity, ftol, &iter, &L_phi, &L_y);
      sprintf(comments," Gen. Gauss: alpha=%.1f, iter=%d", alpha, iter);
      }
/*--------------------------------
* Deconvolution with convex sqrt(s2+x2) regularisation 
* ss is the level which is a kind of (intensity) threshold 
* between "good signal"  and "medium" or "bad" signal 
* (penalty is less severe than tikhonov, for intensities > ss)
*/
  else if(!strncmp(option,"sqrtr",5)) {
      dcv_2D_sqrtrf(alpha, ss, positivity, ftol, &iter, &L_phi, &L_y);
      sprintf(comments," sqrtr: alpha=%.1f, ss=%f iter=%d", alpha, ss, iter);
      }
/*--------------------------------
* Deconvolution with Maximum Entropy method 
*/
  else if(!strncmp(option,"maxen",5)) {
      dcv_2D_mem(alpha, ftol, &iter, &L_phi, &L_y);
      sprintf(comments," MEM: alpha=%.1f, iter=%d", alpha, iter);
      }
/*--------------------------------
* Deconvolution with Gauss Markov regularisation (on x_i+1 - x_i) 
*/
  else if(!strncmp(option,"gmark",5)) {
      dcv_2D_gmark(alpha, positivity, ftol, &iter, &L_phi, &L_y);
      sprintf(comments," Gauss-Markov: alpha=%.1f, iter=%d", alpha, iter);
      }
  else {
      printf(" Fatal: invalid option\n");
      exit(-1);
      }

if(simu) {
      ssum = 0.;
      for(i = 1; i <= nx * ny; i++) ssum += SQUARE(yyd[i] - yy0[i]);
      rms_err = sqrt(ssum)/ (double)(nx*ny);
      printf("rms error=%f\n", rms_err);
      }

/*----------------- L Curve:
* L_phi = phi(x)
* L_y = || y - H x ||^2
*/
if(L_phi != 0.) 
   printf("L curve: L_phi=%.4e L_y=%.4e\n", L_phi, L_y);

sprintf(filename,"%s_yyd",prefix);
RECENT_FFT_DOUBLE(yyd, yyd, &nx, &ny, &nx);
sprintf(comments,"deconvolution with %s and %.1f", option, alpha);
JLP_D_WRITEIMAG(yyd, &nx, &ny, &nx, filename, comments);

dcv_2D_free();
free(hh);
return(0);
}
/**************************************************************
*
**************************************************************/
int dcv_remove_background(double *hh, int nx, int ny, 
                          double *backg, double *noise)
{
register int i;

/* in "auto_scale1.c": */
auto_sky_d(&hh[1], nx, ny, nx, backg, noise);

for(i = 1; i <= nx*ny; i++) hh[i] -= (*backg);

return(0);
}
/****************************************************************
* Routine to select parameters for deconvolution
*
* INPUT:
*  yy[]: signal array
*  nx, ny: size of yy[] 
*  noise: noise measured in the corners of the yy[] array
*
* OUTPUT:
* alpha: regularization parameter 
* ss: used by sqrtr 
* positivity: 1 if positivity constraint
* ftol: tolerance for minimization
****************************************************************/
int dcv_2D_get_param(double *yy, int nx, int ny, double *alpha, 
                     double *ss, int *positivity, double *ftol,
                     double noise, char *option) 
{
register int i;
double maxval, ssum, alpha0, norm_l2;

/* Look for maximum value:
*/
maxval = yy[1];
for( i = 1; i <= nx * ny; i++) if(maxval < yy[i]) maxval = yy[i];

 ssum = 0.;
 for(i = 1; i <= nx * ny; i++) ssum += SQUARE(yy[i]);
 norm_l2 = ssum / (double)(nx * ny);
 noise = MAXI(noise, 1.e-12);

 printf(" Maximum value: %.4e variance: %.4e, noise: %.4e, SNR: %.4e\n",
          maxval, norm_l2, noise, sqrt(norm_l2)/noise);

/* Input of ss needed for sqrtr: */
/* ss is the level which is a kind of (intensity) threshold 
* between "good signal"  and "medium" or "bad" signal 
* (penalty is less severe than tikhonov, for intensities > ss)
*/
   if(!strncmp(option, "sqrtr", 5)) {
      printf(" sqrt(s^2+x^2)/Enter ss value (0.1 or 0.001 for instance):\n"); 
      scanf("%lf", ss);
      }

 alpha0 = 1.;

 if(!strncmp(option, "tikho", 5)) 
    alpha0 = noise / norm_l2;
 else if(!strncmp(option, "ggaus", 5)) 
    alpha0 = noise / pow(norm_l2, 0.55);
 else if(!strncmp(option, "maxen", 5)) 
    alpha0 = noise / (2. * sqrt(norm_l2) * log(norm_l2));
 else if(!strncmp(option, "sqrtr", 5)) 
    alpha0 = noise / sqrt(SQUARE(*ss) + norm_l2);
 else if(!strncmp(option, "gmark", 5)) 
    alpha0 = noise / norm_l2;
/* For wiener: alpha is the ratio variance of noise / variance of signal 
*    alpha = pow(10.,-snr/(double)10.); for simulations where snr is in dB */
 else if(!strncmp(option, "wiene", 5))
    alpha0 = noise / norm_l2;

/* Now input of the other parameters: */
if(!strncmp(option, "tikho", 5) || !strncmp(option, "ggaus", 5) 
   || !strncmp(option, "sqrtr", 5) || !strncmp(option, "gmark", 5)
   || !strncmp(option, "maxen", 5)) { 
     printf(" tolerance ? [e.g., 1.e-5]\n"); 
     scanf("%lf", ftol);
     if(!strncmp(option, "maxen", 5)) 
       *positivity = 1; 
     else 
       {
       printf(" positivity constraint ? [1,0]\n"); 
       scanf("%d", positivity);
       }
     }

   if(!strncmp(option, "wiene", 5)) {
     printf("NB: Wiener filter is 1. / (power spectrum of transfer fct + alpha)\n");
     }

if(!strncmp(option, "tikho", 5) || !strncmp(option, "ggaus", 5) 
   || !strncmp(option, "sqrtr", 5) || !strncmp(option, "gmark", 5)
   || !strncmp(option, "maxen", 5) || !strncmp(option, "wiene", 5))
   {
    printf(" Value for alpha ? (indicative value: %.4e)\n", alpha0);
    scanf("%lf", alpha);
   }

return(0);
}
